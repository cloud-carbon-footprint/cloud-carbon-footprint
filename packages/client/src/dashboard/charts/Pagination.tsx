/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, {
  useState,
  useEffect,
  PropsWithChildren,
  ReactElement,
} from 'react'
import {
  ChevronLeft,
  ChevronRight,
  FirstPage,
  LastPage,
} from '@material-ui/icons'
import { IconButton, makeStyles } from '@material-ui/core'

interface UsePaginateData<T> {
  paginatedData: T[][]
  totalPages: number
}
export const usePaginateData: <T>(
  data: T[],
  pageSize: number,
) => UsePaginateData<T> = (data, pageSize) => {
  const paginatedData = []
  const newEntries = [...data]
  while (newEntries.length > 0) {
    const paginatedSubData = newEntries.splice(0, pageSize)
    paginatedData.push(paginatedSubData)
  }
  return {
    paginatedData,
    totalPages: paginatedData.length,
  }
}

const useStyles = makeStyles(() => {
  return {
    paginationContainer: {
      display: 'flex',
      width: '100%',
      justifyContent: 'flex-end',
      alignItems: 'center',
    },
  }
})

interface PaginateData<T> {
  data: T[]
  pageSize: number
}

export interface Page<T> {
  data: T[]
  page: number
}

export interface PaginationProps<T> extends PaginateData<T> {
  handlePage: (page: Page<T>) => void
}

export const Pagination: <T>(
  props: PropsWithChildren<PaginationProps<T>>,
) => ReactElement = ({ data, pageSize, handlePage }) => {
  const { paginationContainer } = useStyles()
  const [page, setPage] = useState(0)
  const { paginatedData, totalPages } = usePaginateData<typeof data[0]>(
    data,
    pageSize,
  )
  const visibleRows = `${page * pageSize + 1} - ${
    page * pageSize + paginatedData[page]?.length
  }`

  useEffect(() => {
    handlePage({ data: paginatedData[0] || [], page: 0 })
    setPage(0)
  }, [JSON.stringify(data)])

  const onPageChange = (newPage: number) => {
    setPage(newPage)
    handlePage({ data: paginatedData[newPage], page: newPage })
  }

  const lastPage = paginatedData.length - 1
  return data.length === 0 ? (
    <div aria-label="no-pagination-data" />
  ) : (
    <div className={paginationContainer}>
      <span style={{ color: '#ababab', fontWeight: 700, marginRight: '8px' }}>
        {visibleRows} of {data.length}
      </span>
      <IconButton
        color="primary"
        aria-label="first page"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(0)}
      >
        <FirstPage />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-left"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(page - 1)}
      >
        <ChevronLeft />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-right"
        component="button"
        disabled={page === totalPages - 1}
        onClick={() => onPageChange(page + 1)}
      >
        <ChevronRight />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="last page"
        component="button"
        disabled={page === lastPage}
        onClick={() => onPageChange(lastPage)}
      >
        <LastPage />
      </IconButton>
    </div>
  )
}
