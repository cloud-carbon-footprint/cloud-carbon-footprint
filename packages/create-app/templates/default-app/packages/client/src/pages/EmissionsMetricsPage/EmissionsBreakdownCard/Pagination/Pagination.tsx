/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, {
  useState,
  useEffect,
  PropsWithChildren,
  ReactElement,
} from 'react'
import {
  ChevronLeft,
  ChevronRight,
  FirstPage,
  LastPage,
} from '@material-ui/icons'
import { IconButton } from '@material-ui/core'
import { Page } from '../../../../Types'
import useStyles from './paginationStyles'

interface UsePaginateData<T> {
  paginatedData: T[][]
  totalPages: number
}

interface PaginateData<T> {
  data: T[]
  pageSize: number
}

interface PaginationProps<T> extends PaginateData<T> {
  handlePage: (page: Page<T>) => void
}

const usePaginateData: <T>(
  data: T[],
  pageSize: number,
) => UsePaginateData<T> = (data, pageSize) => {
  const paginatedData = []
  const newEntries = [...data]
  while (newEntries.length > 0) {
    const paginatedSubData = newEntries.splice(0, pageSize)
    paginatedData.push(paginatedSubData)
  }
  return {
    paginatedData,
    totalPages: paginatedData.length,
  }
}

const Pagination: <T>(
  props: PropsWithChildren<PaginationProps<T>>,
) => ReactElement = ({ data, pageSize, handlePage }) => {
  const { paginationContainer, paginationLabel } = useStyles()
  const [page, setPage] = useState(0)
  const { paginatedData, totalPages } = usePaginateData<typeof data[0]>(
    data,
    pageSize,
  )
  const lastPage = paginatedData.length - 1
  const visibleRows = `${page * pageSize + 1} - ${
    page * pageSize + paginatedData[page]?.length
  }`

  useEffect(() => {
    handlePage({ data: paginatedData[0] || [], page: 0 })
    setPage(0)
  }, [JSON.stringify(data)])

  const onPageChange = (newPage: number) => {
    setPage(newPage)
    handlePage({ data: paginatedData[newPage], page: newPage })
  }

  if (data.length === 0) {
    return <div aria-label="no-pagination-data" />
  }

  return (
    <div className={paginationContainer}>
      <span className={paginationLabel}>
        {visibleRows} of {data.length}
      </span>
      <IconButton
        color="primary"
        aria-label="first page"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(0)}
      >
        <FirstPage />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-left"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(page - 1)}
      >
        <ChevronLeft />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-right"
        component="button"
        disabled={page === totalPages - 1}
        onClick={() => onPageChange(page + 1)}
      >
        <ChevronRight />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="last page"
        component="button"
        disabled={page === lastPage}
        onClick={() => onPageChange(lastPage)}
      >
        <LastPage />
      </IconButton>
    </div>
  )
}

export default Pagination
