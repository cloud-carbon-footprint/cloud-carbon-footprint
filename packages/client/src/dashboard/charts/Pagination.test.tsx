/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import { create } from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import React from 'react'
import { Pagination } from './Pagination'

describe('Pagination', () => {
  const handlePage = jest.fn()

  const data = [
    { x: 'ebs', y: 3015.014 },
    { x: 'ec2', y: 2521.406 },
    { x: 's3', y: 1718.017 },
    { x: 'lambda', y: 3015.014 },
    { x: 'rds', y: 2521.406 },
    { x: 'msk', y: 1718.017 },
    { x: 'mq', y: 3015.014 },
    { x: 'redshift', y: 2521.406 },
    { x: 'glue', y: 1718.017 },
  ]

  const page1 = [
    { x: 'ebs', y: 3015.014 },
    { x: 'ec2', y: 2521.406 },
    { x: 's3', y: 1718.017 },
  ]

  const page2 = [
    { x: 'lambda', y: 3015.014 },
    { x: 'rds', y: 2521.406 },
    { x: 'msk', y: 1718.017 },
  ]

  const page3 = [
    { x: 'mq', y: 3015.014 },
    { x: 'redshift', y: 2521.406 },
    { x: 'glue', y: 1718.017 },
  ]

  beforeEach(() => {
    handlePage.mockClear()
  })

  it('should render empty div when pagination data is an empty array', () => {
    const { getByLabelText } = render(
      <Pagination data={[]} handlePage={handlePage} pageSize={3} />,
    )
    const noPaginationDataDiv = getByLabelText('no-pagination-data')

    expect(handlePage).toHaveBeenCalledWith({ data: [], page: 0 })
    expect(noPaginationDataDiv).toBeDefined()
  })

  it('renders Pagination with correct configuration', () => {
    const wrapper = create(
      <Pagination data={data} handlePage={handlePage} pageSize={3} />,
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('first page of data should be returned on mount', () => {
    render(<Pagination data={data} handlePage={handlePage} pageSize={3} />)
    expect(handlePage).toHaveBeenLastCalledWith({ data: page1, page: 0 })
  })

  it('previous button should be disabled on first page', () => {
    const { getByLabelText } = render(
      <Pagination data={data} handlePage={handlePage} pageSize={3} />,
    )
    const prevButton = getByLabelText('chevron-left')

    act(() => {
      fireEvent.click(prevButton)
    })

    expect(prevButton).toBeDisabled()
    expect(handlePage).toHaveBeenCalledTimes(1)
  })

  it('should be able to traverse available pages', () => {
    const { getByLabelText } = render(
      <Pagination data={data} handlePage={handlePage} pageSize={3} />,
    )
    const nextButton = getByLabelText('chevron-right')
    const prevButton = getByLabelText('chevron-left')

    expect(handlePage).toHaveBeenLastCalledWith({ data: page1, page: 0 })
    act(() => {
      fireEvent.click(nextButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(2)
    expect(handlePage).toHaveBeenLastCalledWith({ data: page2, page: 1 })

    act(() => {
      fireEvent.click(nextButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(3)
    expect(nextButton).toBeDisabled()
    expect(handlePage).toHaveBeenLastCalledWith({ data: page3, page: 2 })

    act(() => {
      fireEvent.click(nextButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(3)
    expect(nextButton).toBeDisabled()

    act(() => {
      fireEvent.click(prevButton)
    })
    expect(handlePage).toHaveBeenLastCalledWith({ data: page2, page: 1 })
  })

  it('should go to last page on click of last page button', () => {
    const { getByLabelText } = render(
      <Pagination data={data} handlePage={handlePage} pageSize={3} />,
    )
    const lastPageButton = getByLabelText('last page')

    expect(handlePage).toHaveBeenLastCalledWith({ data: page1, page: 0 })
    act(() => {
      fireEvent.click(lastPageButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(2)
    expect(handlePage).toHaveBeenLastCalledWith({ data: page3, page: 2 })

    act(() => {
      fireEvent.click(lastPageButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(2)
    expect(lastPageButton).toBeDisabled()
  })

  it('should go to first page on click of first page button', () => {
    const { getByLabelText } = render(
      <Pagination data={data} handlePage={handlePage} pageSize={3} />,
    )
    const lastPageButton = getByLabelText('last page')
    const firstPageButton = getByLabelText('first page')

    expect(handlePage).toHaveBeenLastCalledWith({ data: page1, page: 0 })
    act(() => {
      fireEvent.click(lastPageButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(2)
    expect(handlePage).toHaveBeenLastCalledWith({ data: page3, page: 2 })

    act(() => {
      fireEvent.click(firstPageButton)
    })
    expect(handlePage).toHaveBeenCalledTimes(3)
    expect(handlePage).toHaveBeenLastCalledWith({ data: page1, page: 0 })
  })
})
