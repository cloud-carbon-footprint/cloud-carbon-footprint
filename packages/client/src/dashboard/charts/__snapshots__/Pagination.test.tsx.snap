// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders Pagination with correct configuration 1`] = `
<div
  className="makeStyles-paginationContainer-2"
>
  <span
    style={
      Object {
        "color": "#ababab",
        "fontWeight": 700,
        "marginRight": "8px",
      }
    }
  >
    1 - 3
     of 
    9
  </span>
  <button
    aria-label="first page"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
        />
      </svg>
    </span>
  </button>
  <button
    aria-label="chevron-left"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
        />
      </svg>
    </span>
  </button>
  <button
    aria-label="chevron-right"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
      </svg>
    </span>
  </button>
  <button
    aria-label="last page"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
        />
      </svg>
    </span>
  </button>
</div>
`;
